lib/mixlib/install/generator/powershell.rb

```diff
@@ -65,14 +65,29 @@ def ps1_modularize(module_body, module_name)
         end
 
         def artifactory_urls
-          artifacts = Mixlib::Install::Backend::Artifactory.new(options).info
-          get_script("get_project_metadata_for_artifactory.ps1", artifacts: artifacts)
+          get_script("get_project_metadata_for_artifactory.ps1",
+                     artifacts: artifactory_backend.info)
+        end
+
+        def artifactory_backend
+          @artifactory_backend ||= Mixlib::Install::Backend::Artifactory.new(options)
+        end
+
+        def product_version
+          if options.for_artifactory?
```

Comment:
To make this better I would move version resolution logic into options @patrick-wright. i.e. 

```
# options.rb
def resolve_version(artifacts)
  if latest?
    @resolved_version ||= begin
      # ...
      # logic to resolve the version
      #
    else
      @resolved_version ||= version
    end
end

# powershell.rb

def artifactory_urls
  artifacts = artifactory_backend.info
  # Above line resolves the version and resolved version is cached in options
  # ^^ Comment is important because we need to call info before doing anything else :)
  
  get_script("get_project_metadata_for_artifactory.ps1",
                  artifacts: artifacts,
                  resolved_version: options.resolved_version)
  # change the template to use `resolved_version` in it.

# Also use options.resolve_version in backend/artifactory.rb
```

Hopefully makes sense. If not we can chat :smile: